File: daily_problems/2024/10/1024/personal_submission/cf195e_yawn_sean.py
# Submission link: https://codeforces.com/contest/305/submission/287591312

mod = 10 ** 9 + 7
class UnionFind:
    def __init__(self, n):
        self.parent_or_size = [-1] * n
        self.depth = [0] * n
 
    def find(self, a):
        acopy = a
        v = 0
        
        while self.parent_or_size[a] >= 0:
            v += self.depth[a]
            if v >= mod: v -= mod
            a = self.parent_or_size[a]
        
        while acopy != a:
            self.depth[acopy], v = v, v - self.depth[acopy]
            if v < 0: v += mod
            self.parent_or_size[acopy], acopy = a, self.parent_or_size[acopy]
        return a
 
    def merge(self, a, b, w):
        pa, pb = self.find(a), self.find(b)
        va, vb = self.getDepth(a), self.getDepth(b)
        
        w += va
        if w >= mod: w -= mod
        w -= vb
        if w < 0: w += mod
        
        self.parent_or_size[pa] += self.parent_or_size[pb]
        self.parent_or_size[pb] = pa
        self.depth[pb] = w
 
    def getDepth(self, a):
        self.find(a)
        return self.depth[a]

def main():
    n = II()
    dsu = UnionFind(n)
    ans = 0

    for i in range(n):
        tmp = LII()
        for j in range(1, len(tmp), 2):
            idx = tmp[j] - 1
            val = tmp[j + 1]
            if val < 0: val += mod
            
            val += dsu.getDepth(idx)
            if val >= mod: val -= mod
            
            ans += val
            if ans >= mod: ans -= mod
            
            u = dsu.find(idx)
            dsu.merge(i, u, val)

    print(ans)

File: daily_problems/2024/10/1024/personal_submission/cf195e_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include <bits/extc++.h>
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std; 

int mod = 1e9 + 7;
struct unionfind {
  public:
    unionfind() : _n(0) {}
    explicit unionfind(int n) : _n(n), parent_or_size(n, -1), depth(n, 0) {}

    void merge(int a, int b, int w) {
        assert(0 <= a && a < _n);
        assert(0 <= b && b < _n);
        int x = search(a), y = search(b);
        int vx = getDepth(a), vy = getDepth(b);

        w += vx;
        if (w >= mod) w -= mod;

        w -= vy;
        if (w < 0) w += mod;

        parent_or_size[x] += parent_or_size[y];
        parent_or_size[y] = x;
        depth[y] = w;
    }

    int search(int a) {
        assert(0 <= a && a < _n);
        int acopy = a, v = 0, tmp;

        while (parent_or_size[a] >= 0) {
            v += depth[a];
            if (v >= mod) v -= mod;
            a = parent_or_size[a];
        }

        while (acopy != a) {
            v -= depth[acopy];
            depth[acopy] += v;
            if (v < 0) v += mod;

            tmp = parent_or_size[acopy];
            parent_or_size[acopy] = a;
            acopy = tmp;
        }
        return a;
    }

    int getDepth(int a) {
        search(a);
        return depth[a];
    }

  private:
    int _n;
    std::vector<int> parent_or_size;
    std::vector<int> depth;
};

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
    
    int n;
    cin >> n;

    unionfind dsu(n);
    int ans = 0;

    for (int i = 0; i < n; i ++) {
        int k;
        cin >> k;

        while (k --) {
            int idx, val;
            cin >> idx >> val;
            idx --;

            if (val < 0) val += mod;
            val += dsu.getDepth(idx);
            if (val >= mod) val -= mod;

            ans += val;
            if (ans >= mod) ans -= mod;

            int u = dsu.search(idx);
            dsu.merge(i, u, val);
        }
    }

    cout << ans;

    return 0;
}